Fix nontermination with `-short-paths` and recursive modules (#14036)

brandonzstride opened the pull request:
This pull request proposes a potential solution to #14036.

Disclaimer: I am not versed in the OCaml compiler internals. This PR contains a possibly suboptimal fix, and I discuss what I might have done if I had more experience with the compiler.

When resolving short paths for certain cyclic type abbreviations in recursive modules, `Out_type.normalize_type_path` enters a loop because the typing environment contains a cycle of paths. A similar issue was fixed in #12645, which abstracted the environment to prevent nonterminating resolution of most other cyclic type definitions. That approach, however, does not immediately extend to this case.

## Proposed change

I propose detecting the cycle directly where the loop occurs. The change is localized to `typing/out_type.ml`.

## Performance

Since this logic only runs when printing with `-short-paths`, and both cycles and paths are expected to be small, I expect the performance impact of this PR is negligible.

## Tests

I've added tests at `testsuite/tests/typing-short-paths/cycles.ml`, including the example from #14036 and several new related cases.

Currently, test failure is only detected at the 10-minute timeout, which is excessive--just a few seconds would suffice. I don't know how to configure a shorter timeout in `ocamltest`. I recommend caution in merging these tests until the timeout issue is addressed.

## Implementation concerns

The changes might be at odds with the intention of `normalize_type_path`, which is expected to evaluate to a normalized or canonical type path. Here, an arbitrary path is chosen from the cycle, which may not be ideal. Other parts of the code rely on `normalize_type_path` to check for path ambiguity, which this change may affect. That said, no existing tests fail, so it may be safe in practice. Further, `normalize_type_path` was previously a partial function, and the inputs affected by this PR were already undefined. This patch makes its behavior in those cases defined and terminating.

In #12645, the issue was addressed with a more principled approach by printing paths from an abstract environment. However, when checking recursive modules, the necessary information to construct such an abstract environment is not readily available, so the concrete environment with the physical cycle is used (see `check_recmod_typedecl` at `typedecl.ml`, lines 1916-1924, and its use at `typemod.ml`, lines 200-209). Ideally, an abstract environment would be used here as well, but in this PR I settle for a localized fix.

While this patch resolves the issue in a localized way and causes no regressions in any existing test, it may not be the maintainers' preferred fix.

Octachron commented:
Thanks for the bug report and the patch!

However, a suitably abstract environment is in fact available when typechecking recursive modules. You can modify the `Typedecl.check_recmod_typedecl` function to take an abstract environement as argument, and then use the initial environment for typechecking the recursive module group as an abstract environment.

brandonzstride commented:
Thank you for the help. I appreciate your patience, too, as I took some time to get back to this.

I reverted the change to `Out_type` and made the appropriate changes to `Typedecl` and `Typemod` so that recursive modules pass the original, abstracted environment as the printing environment. As you expected, this fully fixes the issue.

Some tests changed accordingly, and I think the new behavior is favorable. My concerns about timeout as a failure mode still stand, though.

Please let me know if it seems I've made a mistake here, or if this isn't what you had in mind.

Octachron commented on typing/typemod.ml:
```diff
@@ -1888,7 +1888,7 @@ and transl_recmodule_modtypes env sdecls =
   in
   let env1 = make_env dcl1 in
-  check_recmod_typedecls env1 (map_mtys dcl1);
+  check_recmod_typedecls ~abs_env:env env1 (map_mtys dcl1);
```
The environment `env` is in fact not the best environment because it does not contain any of the contents of the current group of recursive modules. The abstract environment is in fact the initial environment `env0`, where the type definitions have been abstracted. One example where it might make a difference in the future would be

```ocaml
module rec X : sig
  type 'a t = 'a X.s as 'b
    constraint 'b = int
  type 'a s = int
end = X
```
which currently emits the following error message

>``` 
>Error: The type constraints are not consistent.
>       Type "'a X.s" is not compatible with type "int"
>```

but by using `env0` (possibly renamed to `abs_env`) and tweaking the function `abstract_type_decl` to give a better reason on why the type declaration was abstracted, we could extend the error message to
>```
> Error: The type constraints are not consistent.
>       Type "'a X.s" is not compatible with type "int"
>       Type "X.s" was considered abstract when checking constraints in this
>       recursive module definition.
>```

brandonzstride replied:
I added an origin to the `Types` module for abstract types that are for approximating recursive modules signatures, and then I was able to add this error message.

Octachron commented on typing/typemod.ml:
```diff
@@ -385,7 +385,7 @@ let check_well_formed_module env loc context mty =
           let (id_mty_l, rem) = extract_next_modules rem in
           begin try
-            check_recmod_typedecls (Lazy.force env) ((id, mty) :: id_mty_l)
+            check_recmod_typedecls ~abs_env:orig_env (Lazy.force env) ((id, mty) :: id_mty_l)
```
This unfortunately incorrect: the environment `orig_env` does not contain any of the contented of the functor body that we are checking. Thus with this change, we end up rejecting

```ocaml
module Constraint(F:sig type 'a t end-> sig type 'a t end) = struct
  type 'a x = 'b constraint 'a = 'b * 'b
  module rec Fixed: sig
    type 'a s = < x: 'a F(Fixed).t >
    type 'a t = ('a  * 'a) x s
  end = Fixed
end
module Ok = Constraint(functor (X:sig type 'a t end) -> X)
``` 
because we end up checking the regularity of the type `'a s` in an environment where we don't know that `('a * 'a) x = 'a`.

The correct fix here is to construct an abstract environement from the signature `sg` by making type declarations abstract only in the recursive group being checked. 

Since the fix is quite more involved, I would propose to split this fix in an ulterior PR and use `env` as an `abs_env` temporarily. And if this ulterior fix is too much work for you, I will take care of this part.

brandonzstride replied:
I put in the temporary patch to use the forced `env`. The more involved fix will be too much for me, as it stands.

smuenzel replied:
The more involved fix is in progress: https://github.com/smuenzel/ocaml/pull/1

Octachron commented on testsuite/tests/typing-short-paths/cycles.ml:
```diff
@@ -0,0 +1,116 @@
+         "('a, 'b, bool) Y.t" = "('b, 'a) X.t",
+         "('b, 'a) X.t" = "('a, 'b, bool) Y.t",
+         "('a, 'b, bool) Y.t" = "('b, 'a) X.t"
+|}]
```
You probably want to add a case for the `check_recmod_typedecl` function:

```ocaml
type a__name__that__shall__not__be__printed

module type T = sig type t end
module Fix(F: T -> T) = struct
  module Atlas = struct type t = a__name__that__shall__not__be__printed end
  module rec Fixed: sig
    type t = F(Fixed).t
  end = F(Fixed)
end
module Err = Fix(functor (X:T) -> struct type t = a__name__that__shall__not__be__printed -> X.t end)
```


and the constraint example:

```ocaml
module Constraint(F:sig type 'a t end-> sig type 'a t end) = struct
  type 'a x = 'b constraint 'a = 'b * 'b
  module rec Fixed: sig
    type 'a s = < x: 'a F(Fixed).t >
    type 'a t = ('a  * 'a) x s
  end = Fixed
end
[%%expect {|
module Constraint :
  (F : sig type 'a t end -> sig type 'a t end) ->
    sig
      type 'a x = 'b constraint 'a = 'b * 'b
      module rec Fixed :
        sig type 'a s = < x : 'a F(Fixed).t > type 'a t = ('a * 'a) x s end
    end
|}]

module Ok = Constraint(functor (X:sig type 'a t end) -> X)
```

Octachron commented on testsuite/tests/typing-short-paths/cycles.ml:
```diff
@@ -0,0 +1,116 @@
+(* TEST
+ flags = " -short-paths ";
+ expect;
```
In order to avoid the long timeout on failure, you can limit the stack size by adding
```
ocamlrunparam +="l=100000";
```

Octachron commented:
Thanks a lot for the prompt update! Don't feel pressured in term of time. 

There is still a small tweak needed for the recursive module case.

The fix for the check on functor application is more complex, but it might be a good idea to split into a second PR, depending on how much time you wish to spend on the PR.

gasche commented:
Gentle ping. @brandonzstride , would you be available to push this past the finish line? If not, please feel free to say so, we can probably rebase and tweak the PR on our side.

brandonzstride commented:
@gasche Thank you for the ping. I'm sorry for letting this hang for so long. I feel bad about that.

I'd like to take another shot at this if you feel you can afford to give me the learning opportunity here. Now is actually a fine time for me to revisit this.

brandonzstride commented:
Rebased onto trunk.

brandonzstride commented:
I made the simple fixes that you suggested, and I added the tests that my previous commits caused to fail. Those examples can type check now, but I don't add more examples of the same spirit.

I also pushed an attempt to add the error message to explain why types are inconsistent due to abstracting them while checking recursive modules. This change touched quite a few files, but the changes are all small.

Octachron commented on typing/typemod.ml:
```diff
@@ -417,7 +417,7 @@ let rec extract_next_modules = function
   | sg -> ([], sg)
 
-let check_well_formed_module env loc context mty =
+let check_well_formed_module orig_env loc context mty =
```
I would propose to keep the name `env` in this PR to reduce noise.

brandonzstride replied:
Thanks. I had forgotten that I had made it `orig_env` for my previous, incorrect change. Reverted.

Octachron commented on typing/typedecl.ml:
```diff
@@ -1911,24 +1911,24 @@ let abstract_type_decl ~injective arity =
   end
 
-let approx_type_decl sdecl_list =
+let approx_type_decl sdecl_list ~abstract_origin =
```
Nitpicking: as `abstract_origin` is an auxillary parameter, I would propose to move it before the main argument `sdecl_list` (while renaming it to fit your choice for `abstract_type_decl`).

Octachron commented on typing/typedecl.ml:
```diff
@@ -1890,13 +1890,13 @@ let transl_package_constraint ~loc env ty =
 (* Approximate a type declaration: just make all types abstract *)
 
-let abstract_type_decl ~injective arity =
+let abstract_type_decl ~injective ~abstract_origin arity =
```
I would say this is more an `abstract_reason` or maybe just `reason` or `explanation`  rather than an origin.

brandonzstride replied:
To me, `reason` or `explanation` might suggest that `abstract_type_decl` branches on it--that it does something differently because of this parameter's value. I think `abstract_origin` helps imply that it is simply inserted as data. It is a data parameter, not a behavior-guiding parameter, and I hope the name can convey that. Is there a convention I'm missing about how parameters are typically named?

Octachron replied:
I believe it might be the implicit context that `explanation`s are metadata added for displaying information to human users (in error messages) and rarely used to change behavior.

brandonzstride replied:
Got it. This convention makes sense to me. I went with `explanation`, then. I did not update the documentation comment because anything I try to write that further ties "explanation" with "origin" just feels redundant.

Octachron commented on typing/typedecl.mli:
```diff
@@ -47,12 +47,18 @@ val transl_with_constraint:
+    Parsetree.type_declaration list -> abstract_origin:Types.type_origin ->
                                   (Ident.t * type_declaration) list
+
+(* [abs_env] is an abstract environment without physical cycles.
```
As a documentation comment, this should use `(**`.

Octachron commented on typing/typedecl.mli:
```diff
@@ -47,12 +47,18 @@ val transl_with_constraint:
+
+(* [abs_env] is an abstract environment without physical cycles.
+   It is used as a printing environment in the case of cycles.
+   The other environment has already tied that knot. *)
```
As a documentation comment, this should use `(**`. You can probably add the description existing on the ml side,
and expand the last sentence which might be obscure outside of the context of this PR:

```ocaml
(** [check_recmod_typedecl ~abs_env env loc recmod_ids path decl] 
   -  [recmod_ids] is the list of recursively-defined module idents.
   - [path, decl] is the type declaration to be checked. 
   - [abs_env] is an abstract environment without physical cycles. It is used as a printing environment.
   - [env] : the main environment, which may contain cycles introduced by the recursive module
      definitions.
*)
```

Octachron commented on typing/typedecl.mli:
```diff
@@ -47,12 +47,18 @@ val transl_with_constraint:
-val abstract_type_decl: injective:bool -> int -> type_declaration
+val abstract_type_decl:
+    injective:bool -> abstract_origin:Types.type_origin -> int -> type_declaration
+(* Approximate a type declaration with abstract types of the given origin. *)
```
This is a documentation comment, thus  `(**` rather than `(*`. 

> * Approximate a type declaration as abstract types of the given origin.

Octachron reviewed (approved):
The changes look good now: I am left with only documentation and nitpicking comments (that you can ignore if you wish).

You also should add a Changelog entry in the bug fix section.

brandonzstride commented:
Thanks for all the feedback! I made the trivial changes, and I'm happy to stick around through the rest of the nitpicking until I get this right.

I updated the Changelog, but it feels like combining two in one: we have the bug fix for nontermination of recursive module cycle printing, and we also have the small addition to error messages.

Octachron commented on typing/typedecl.ml:
```diff
@@ -1911,24 +1911,28 @@ let abstract_type_decl ~injective arity =
+(* [abs_env] is an abstract environment without physical cycles.
+  It is used as a printing environment in the case of cycles.
+  [env] has already tied that knot, and it is the main typing
+  environment, which may contain cycles. *)
```
The "tied that knot" part and the "which may contain cycles" are redundant, you should keep only one of them.

Octachron commented:
There are some remaining rogue newlines and overly long lines, you can have a look at https://github.com/ocaml/ocaml/blob/trunk/HACKING.adoc#check-typo to see how to run the style linter locally.

Note that I am planning to squash the PR into one commit once `check-typo` is satisfied to avoid the back-and-forth between the difference implementations.

brandonzstride commented:
Sorry that I wasn't running the linter originally. I think I just fixed all the style issues.

And that's understandable to squash it all into one commit because I had to undo so many changes from earlier commits. Fine by me. Thanks for all the help and for letting me work through this! Please let me know if there's anything else I should do.

Octachron commented:
Hm, in fact the error improvement part is a cleanly separated part of the PR. I propose to rewrite the history to have a commit for the main bug fix and another for the error message part. I can do this part if you prefer.

brandonzstride committed:
Use abstract environment to print recmod paths (#14036)

The printing of type cycles with -short-paths previously did not
terminate because it used a typing environment with the physical
cycles. Now, an abstract typing environment is used to print the
types for error messages.

brandonzstride committed:
Add error message for abstract types in recmod signatures

This commit adds the origin of abstract types when checking recursive
module signatures. When errors arise due to type checking only with an
abstract module signature, an error message explains why the type was
seen as abstract.

brandonzstride commented:
I went ahead and rewrote history into two the separated commits. Each builds, passes tests, and passes the linter individually. But thanks for offering to do it!

No rush at all, but when you have a chance, I'd appreciate if you take one more look to make sure it was split as you expected. Thanks.

Octachron commented:
The split in the two commits was exactly what I had in mind. Merged, thanks for your contribution !

brandonzstride commented:
Thank you for the help and guidance! I appreciate you spending the time to help me get this PR right. It was a pleasure.
